Push notification server specs (#139)
status-im/specs: docs/raw/push-notification-server.md

---
permalink: /spec/16
parent: Stable specs
title: 16/PUSH-NOTIFICATION-SERVER
---

# 16/PUSH-NOTIFICATION-SERVER

> Version: 0.1
>
> Status: Raw
>
> Authors: Andrea Maria Piana <andreap@status.im>

> - [Push Notification Server](#16push-notification-server)

Reviewer: I wonder if it makes sense to have a basic components section, so it is clear what the pieces are. E.g.

- Central PN server
- PN service
- Alice Device-ID session
- Bob Device-ID session

Can e..g be a graphviz diagram to showcase how registering works, etc.  E.g. https://specs.vac.dev/specs/remote-log.html#flow

Reviewer 2: Agree. Moving the terminology and high-level flow descriptions upfront followed by the detailed protocol and data structures will help. It will also help to clarify the relationship/mapping among device/client/user/keys.

  - [Reason](#reason)
  - [Requirements](#requirements)
  - [Registering with the push notification service](#registering-with-the-push-notification-server)
  - [Re-registering with the push notification service](#re-registering-with-the-push-notification-server)
  - [Changing options](#changing-options)
  - [Unregistering from push notifications](#unregistering-from-push-notifications)
  - [Advertising a push notification server](#advertising-a-push-notification-server)

>   - [Discovering a push notification server](#discovering-a-push-notificatino-server)

Reviewer 3: ```suggestion
  - [Discovering a push notification server](#discovering-a-push-notification-server)
```

  - [Querying the push notification service](#querying-the-push-notification-server)
  - [Sending a push notification](#sending-a-push-notification)
  - [Flow](#flow)
    - [Registration process](#registration-process)
    - [Sending a notification](#sending-a-notification)

>     - [Receiving a notification](#receiving-a-notification)

Reviewer 3: ```suggestion
    - [Receiving a push notification](#receiving-a-push-notification)
```

  - [Protobuf description](#protobuf-description)
    - [PushNotificationRegister](#pushnotificationregister)
    - [PushNotificationPreferences](#pushnotificationpreferences)
    - [PushNotificationDeviceToken](#pushnotificationdevicetoken)
    - [PushNotificationOptions](#pushnotificationoptions)
    - [PushNotificationPreferencesAcknowledgment](#pushnotificationpreferencesacknowledgement)
    - [ContactCodeAdvertisement](#contactcodeadvertisement)
    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
    - [PushNotificationQuery](#pushnotificationquery)
    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
    - [PushNotification](#pushnotification)
    - [PushNotificationRequest](#pushnotificationrequest)
    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
  - [Partitioned topic](#partitioned-topic)
  - [FAQ](#faq)
  - [Changelog](#changelog)
    - [Version 0.1](#version-01)
  - [Copyright](#copyright)

## Reason

> Push notifications for IOS devices and for some android device can only be

Reviewer: Comment for second round: is this strictly true? in practice this is the conclusion. I'd perhaps elaborate a bit more remote and local options here and why remote for iOS makes sense, and that this implies using the APN service

Author: well depends on your definition of push notification, if you want them to be instant (which is the common understanding), then yes, as we are not allowed to keep a connection open (only voip applications do), and the app would be brought to the foreground to check for messages at intervals, and only when apple thinks is energy efficient (on wifi, when the user switches application etc), and would not be brought back if force quit.

I can add a section to expand on the rationale.

Reviewer: I agree with the conclusion, just that the decision space pruning should come first, i.e. can only do remote or local, local does'nt work due to X. If X were were to change in future (unlikely but hey), it is clear to reader that architecture could change.

implemented relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for IOS or [Firebase](https://firebase.google.com/) for 
Android devices that do not support background services or that often kill the 
background service.
Therefore Status provides a set of Push notification services that can be used
to achieve this functionality.

Because this can't be safely implemented in a privacy preserving manner, clients
MUST be given an option not to enable push notifications and not to send push notifications.

## Requirements

> Status needs to run their own [gorush](https://github.com/appleboy/gorush) 

Reviewer: Second round: the party releasing the app. Which could be Status Gmbh (or PTE LTD?), or some other party.

publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.

## Registering with the push notification service

> A client MAY register with a Push Notification service of their choice.

Reviewer: can they register / get replies from multiple? (partially answered below)

Reviewer: answered below

Author: Yes

Reviewer 4: Maybe make this sentence more clear that multiple PN services can be registered to.

```suggestion
A client MAY register with one or more Push Notification services of their choice.
```

[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)

A message MUST be sent to the [partitioned topic](#partitioned-topic) for the
public key of the node, encrypted with this key.

The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):

```protobuf

message PushNotificationOptions {
  boolean enabled = 1;
  repeated string allowed_user_list = 2;

>   repeated string block_user_list = 3;

Reviewer: Is this supposed to be appending? Could this get out of hand in terms of the number of users allowed/blocked? (I suppose that can be taken care of later)

Author: This is straight replacing, given that this is a fairly rare message and we have a 1MB limit, I think we are safe in terms of size, if size will be a concern we can always implement a partial update.

  repeated string block_chat_list = 4;
}

message PushNotificationDeviceToken {
  enum TokenType {
    UNKWOWN_TOKEN_TYPE = 0;
    APS_TOKEN = 1;
    FIREBASE_TOKEN = 2;
  }
  TokenType token_type = 1;
  string token = 2;
  string installation_id = 3;
  PushNotificationPreferences preferences = 4;
}

message PushNotificationPreferences {

>   repeated DeviceToken device_tokens = 1;

Reviewer: Outside of DeviceToken, is the unique identifier here DeviceToken+ID? How do we deal with the 1-device-N-user case here?

Author: > 1-device-N-user

Not sure I understand exactly the question, you mean multiple users using the same device? Which is essentially, how do we handle multiple accounts on a single installation?

If that's the case, is `installation_id`. Installation id is different for each generated account, so if you have 3 accounts on your device, you will have 3 different installation ids. this is passed to the push notification, so on opening the notification, we can already present the user with the right account and it can type the password.

hope I understood the question correctly.

Reviewer: Multiple accounts/users on one device, correct.

}

message PushNotificationRegister {
  bytes payload = 1;
  bytes signature = 2;
}
```

A push notification server will handle the message according to the following rules:

- it MUST verify that the `signature` matches the public key of the sender.
- it MUST verify that `token_type` is supported
- it MUST verify that `token` is non empty
- it MUST verify that `installation_id` is non empty
- it MUST verify that `version` is non-zero and greater than the currently stored, if any
- it MUST verify that `device_tokens` is non empty

If `signature` does not match the public key of the sender, the message MUST be discarded.

If `token_type` is not supported, a response MUST be sent with `error` set to 
`UNSUPPORTED_TOKEN_TYPE`.

If `token`,`installation_id`,`device_tokens`,`version` are empty, a response MUST
be sent with `error` set to `MALFORMED_MESSAGE`.

If the `version` is equal or less than the currently stored version, a response MUST
be sent with `error` set to `VERSION_MISMATCH`.

If any other error occurs the `error` should be set to `INTERNAL_ERROR`.

If the request is successful, an access token MUST be generated and sent back to the client
with `success` set to `true`.

> It is RECOMMENDED to use a randomly generated string of at least 36 characters.

Reviewer: for which field?

Author: I lost a sentence somewhere :) it's for the access token

Otherwise a response MUST be sent with `success` set to `false`.

`request_id` should be set to the `SHA3` of the signature sent by the client.

The request MUST be sent on the [partitioned topic][./2-account.md] of the sender.

The payload of the response is:

```protobuf

> message PushNotificationPreferencesAcknowledgement {

Reviewer: is this an Ack or more of a Reply? If a generic Ack, any reason we couldn't use data sync for this? Reading fields, it seems more like a Reply to me.

Author: Reply is a more fitting name

Author: I settled on Response

  boolean success = 1;
  ErrorType error = 2;
  bytes request_id = 3;
  PushNotificationPreferences preferences = 4;
  string access_token = 5;

  enum ErrorType {
    UNKNOWN_ERROR_TYPE = 0;
    MALFORMED_MESSAGE = 1;
    VERSION_MISMATCH = 2;
    UNSUPPORTED_TOKEN_TYPE = 3;
    INTERNAL_ERROR = 4;
  }
}
```

[//]: (We can ratched and use a partitioned topic here)

A client SHOULD listen for an acknowledgement sent on their [partitioned topic][./2-account.md].

If `success` is `true` the client has registered successfully.

If `success` is `false`:

- If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried without ensuring 
  that is correctly formed.
- If `INTERNAL_ERROR` is returned, the request MAY be retried, but the client MUST be 
  backoff exponentially
- If `VERSION_MISMATCH` is returned, the client SHOULD ensure that the remote version returned in `preferences` is integrated, and the version should be incremented and the request MAY be sent again.

> A client MAY register with multiple Push Notification Servers in order to increase availability.

Reviewer: I see above is answered here

A client SHOULD make sure that all the notification services they registered with have the same information about their devices and tokens.

If no response is returned the request SHOULD be considered failed and MAY be retried with the same server or a different one, but it MUST exponentially backoff after each trial.

If the request is successful a token is returned. This SHOULD be 
[advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.

## Re-registering with the push notification server

A client SHOULD re-register with the node if the APS or FIREBASE token changes.

When re-registering a client SHOULD ensure that it has the most up-to-date
`PushNotificationPreferences`, update the part relative to their `installation_id`
if necessary, increment `version` and send a `PushNotificationRegister` as described above.

Once re-registered, a client SHOULD advertise the changes.

## Changing options

This is handled in exactly the same way as re-registering above.

## Unregistering from push notifications

> To unregister a client MUST send a `PushNotificationRegister` request as described

Reviewer: Should this be explicit in terms of 'when a user logs out' the client should deregister?

I suppose that is part of a the main client spec, not here

Author: Though is not on logout (otherwise you will stop receiving notifications), you will want to unregister if you don't want to receive notifications anymore, or don't want to use this server anymore.

Reviewer: Hm, it's not clear to me what the correct design is here. If you logout as one user and login as a new user, you could argue that the first user shouldn't receive notifications anymore. To me that might be more intuitive, than getting PNs for some user user.

Use case: family sharing a phone, or some kind of kiosk mode where you shouldn't receive PNs for other suers.

Author: I double checked with UX and indeed is how they think it should be behave, until explicit logout you should receive notifications.
It is a bit of a UX concern, so not sure if it's needed to be specified here.
For now I will add it as a MAY and the implementation in status-react will reflect this behavior.

Reviewer: Fair, agre re deferring it.

@hesterbruikman IMO this UX decision should be revisited. It goes against expectations and also explicitly against previous PN designs I have made, and just sounds off to me. If you are logged off you should NOT receive PNs IMO, that's what that action does. I doubt it is how other IMs work either tbh.

Reviewer 2: This might also be relevant if/when we enable multiple chat keys/identities for the same user. Switching chat keys by the user should be treated the same as user logout above.

above with `unregister` in `PushNotificationPreferences` set to `true`, or removing
their device information.

The server MUST remove all data about this user if `unregistering` is `true`,
apart from the `hash` of the public key and the `version` of the last options,
in order to make sure that old messages are not processed.

## Advertising a push notification server

> Each user registered with one or more push notification servers SHOULD

Reviewer: Just to make sure I understand correctly: Is the idea then that each user who wants to send PNs to that user would pick this up and store it locally?

Author: Yes, we could have a discovery for which push notification server is used by a given user, that's simple:

Every PN server listen to topic(pk-of-user) and the client can query it and the pk of the server(s) is returned, but really offers no benefits at the moment compare to just querying a mailserver, so left it out for simplicity.

When sending a message, a list of `installation_id` for the devices targeted 
by the message SHOULD be collected.

If a message is for all the user devices, all the `installation_id` known to the client MAY be used. 

The number of devices MAY be capped in order to reduce resource consumption.

At least 3 devices SHOULD be targeted, ordered by last activity.

For any device that a token is available, or that a token is successfully queried,
a push notification message SHOULD be sent to the corresponding push notification server.

```protobuf
message PushNotification {
  string access_token = 1;
  string chat_id = 2;
}

message PushNotificationRequest {
  repeated PushNotification requests = 1;
  bytes message = 2;
  string message_id = 3;
  string ack_required = 4;
}
```

Where `message` is the encrypted payload of the message and `chat_id` is the
`SHA3` of the `chat_id`.
`message_id` is the id of the message (link)

If multiple server are available for a given push notification, only one notification
MUST be sent.
In such cases `ack_required` MAY bet set and the sender SHOULD be listening
on the topic derived from the first 4 bytes of `SHA3(ack_required)` and with

> waku encryption key of `ack_required`.

Reviewer: elaborate on waku encryption key here

who requested the token but not necessarily who sent a push notification. 
Correlation between the two can be trivial in some cases.

This also allows a mode of use as we had before, where the server does not propagate
info at all, and it's left to the user to propagate the token, through contact requests
for example.

### Why so many different protobuf messages?

Many are just wrappers, I have not re-used any for now for clarity.

### Why advertise with the bundle?

Advertising with the bundle allows us to piggy-back on an already implemented behavior
and save some bandwidth in cases where is not filtering by public keys 

### What's the  bandwidth impact for this?

Generally speaking, for each 1-to-1 message and group chat message you will sending
1 and `number of participants` push notifications. This can be optimized if 
multiple users are using the same push notification server. Queries have also 
a bandwidth impact but they are made only when actually needed

### What's the information disclosed?

The data disclosed with each message sent by the client is above, but for a summary:

When you register with a push notification service you may disclose:

1) Your public key

> 2) Which devices you have

Reviewer: can you choose to disclose only some of those devices? for whatever reason

Author: Yes, you can always decide not to register with a notification server, or partially register etc

3) The hash of the chat_ids you want to filter out
4) The has of the public keys you are interested/not interested in

When you query a notification service you may disclose:

1) Your public key
2) The fact that you are interested in sending push notification to a given user

Effectively this is fairly revealing if the user has a whitelist implemented.
Therefore sending notification should be optional.

### What prevents a user from generating a random key and getting an access token and spamming?

Nothing really, that's the same as the status app as a whole. the only mechanism that prevents
this is using a whilelist as described above, but that implies disclosing your true identity to the push notification server.

> ### Why not 0-knowledge proofs/quantum computing

Reviewer: lol

    - [PushNotificationPreferencesResponse](#pushnotificationpreferencesresponse)
    - [ContactCodeAdvertisement](#contactcodeadvertisement)
    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
    - [PushNotificationQuery](#pushnotificationquery)
    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
    - [PushNotification](#pushnotification)
    - [PushNotificationRequest](#pushnotificationrequest)
    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
  - [Partitioned topic](#partitioned-topic)
  - [Anonymous mode of operations](#anonymous-mode-of-operations)
  - [FAQ](#faq)
  - [Changelog](#changelog)
    - [Version 0.1](#version-01)
  - [Copyright](#copyright)

## Reason

Push notifications for IOS devices and for some android device can only be
implemented relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for IOS or [Firebase](https://firebase.google.com/).

> This is useful for some Android devices that do not support background services or that often kill the 

Reviewer 3: ```suggestion
This is useful for Android devices that do not support background services or that often kill the 
```

background service.

> IOS only allows certain kind of applications to keep a connection open when in the

Reviewer 3: ```suggestion
iOS only allows certain kind of applications to keep a connection open when in the
```

background, VoIP for example, which current status client does not qualify for.

> Application on IOS can also request execution time when are in the [background]

Reviewer 3: ```suggestion
Applications on iOS can also request execution time when they are in the [background]
```

- it MUST verify that the `signature` matches the public key of the sender.
- it MUST verify that `token_type` is supported
- it MUST verify that `token` is non empty
- it MUST verify that `installation_id` is non empty
- it MUST verify that `version` is non-zero and greater than the currently stored, if any
- it MUST verify that `device_tokens` is non empty

If `signature` does not match the public key of the sender, the message MUST be discarded.

If `token_type` is not supported, a response MUST be sent with `error` set to 
`UNSUPPORTED_TOKEN_TYPE`.

If `token`,`installation_id`,`device_tokens`,`version` are empty, a response MUST
be sent with `error` set to `MALFORMED_MESSAGE`.

If the `version` is equal or less than the currently stored version, a response MUST
be sent with `error` set to `VERSION_MISMATCH`.

If any other error occurs the `error` should be set to `INTERNAL_ERROR`.

If the request is successful, an access token MUST be generated and sent back to the client
with `success` set to `true`.

It is RECOMMENDED to use a randomly generated string for `access_token` of at least 36 characters.

> Otherwise a response MUST be sent with `success` set to `false`.

Reviewer 3: Should this line not be moved up? it seems out of context with the above line of the access token.

---
permalink: /spec/16
parent: Stable specs
title: 16/PUSH-NOTIFICATION-SERVER
---

# 16/PUSH-NOTIFICATION-SERVER

> Version: 0.1
>
> Status: Raw
>
> Authors: Andrea Maria Piana <andreap@status.im>

- [Push Notification Server](#16push-notification-server)
  - [Reason](#reason)
  - [Requirements](#requirements)
  - [Components](#components)

>   - [Registering with the push notification service](#registering-with-the-push-notification-server)

Reviewer 3: ```suggestion
  - [Registering with the push notification service](#registering-with-the-push-notification-service)
```

  - [Protobuf description](#protobuf-description)
    - [PushNotificationRegister](#pushnotificationregister)
    - [PushNotificationPreferences](#pushnotificationpreferences)
    - [PushNotificationDeviceToken](#pushnotificationdevicetoken)
    - [PushNotificationOptions](#pushnotificationoptions)
    - [PushNotificationPreferencesResponse](#pushnotificationpreferencesresponse)
    - [ContactCodeAdvertisement](#contactcodeadvertisement)
    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
    - [PushNotificationQuery](#pushnotificationquery)
    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
    - [PushNotification](#pushnotification)
    - [PushNotificationRequest](#pushnotificationrequest)
    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
  - [Partitioned topic](#partitioned-topic)
  - [Anonymous mode of operations](#anonymous-mode-of-operations)
  - [Security considerations](#security-considerations)
  - [FAQ](#faq)
  - [Changelog](#changelog)
    - [Version 0.1](#version-01)
  - [Copyright](#copyright)

## Reason

> Push notifications for iOS devices and for some Android devices can only be

Reviewer 3: ```suggestion
Push notifications for iOS devices and some Android devices can only be
```

implemented by relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for iOS or [Firebase](https://firebase.google.com/).

> This is useful for Android devices that do not support background services or that often kill the 

Reviewer 2: It will help to mention which are these Android devices? Does it depend on the OEM or OS or something else? Also, my understanding is that our current Android "local" notifications depend on [foreground service](https://github.com/status-im/specs/blob/b31ce655b99580c3e5d10c81207ae829e561757e/docs/draft/notifications.md). So what does support for background services have to do with that? Perhaps they're related. Good to clarify.

Author: my bad, it's foreground service,
in terms of getting killed, that's something I have been told by devs who have looked into this more carefully https://stackoverflow.com/questions/49637967/minimal-android-foreground-service-killed-on-high-end-phone @yenda might have more info

background service.

> iOS only allows certain kind of applications to keep a connection open when in the

Reviewer 2: Interesting. Any pointer to how an application can qualify? Does this mean that if/when Status adds any of such features, it will qualify for this and we will then not need the proposed solution?

Author: Currently it's only for VoIP application I believe

background, VoIP for example, which current status client does not qualify for.

Applications on iOS can also request execution time when they are in the [background](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)
but it has a limited set of use cases, for example it won't schedule any time
if the application was force quit, and generally is not responsive enough to implement
a push notification system.

Therefore Status provides a set of Push notification services that can be used
to achieve this functionality.

Because this can't be safely implemented in a privacy preserving manner, clients

> MUST be given an option not to enable push notifications and not to send push notifications.

Reviewer 2: ~MUST be given an option not to enable push notifications and not to send push notifications.~
How about this:
"MUST be given an option to opt-in to receiving and sending push notifications. They are disabled by default."

## Requirements

> The party releasing the app needs to run their own [gorush](https://github.com/appleboy/gorush) 

Reviewer 2: ~The party releasing the app needs to run their own~
The party releasing the app, Status in this case, needs to run its own

publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.

## Components

### Gorush instance

A [gorush](https://github.com/appleboy/gorush) instance MUST be publicly 
available, this will be used only by push notification servers.

### Push notification server

A push notification server used by clients to register for push notification and

> sending push notifications.

Reviewer 2: ~A push notification server used by clients to register for push notification and sending push notifications.~
A push notification server used by clients to register for receiving and sending push notifications.

### Registering client

> A status client that wants to receive push notifications

Reviewer 2: ~status~ Status

### Sending client

> A status client that wants to send push notifications

Reviewer 2: ~status~ Status

## Registering with the push notification service

A client MAY register with one or more Push Notification services of their choice.

[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)

A `PNR message` (Push Notification Registration) MUST be sent to the [partitioned topic](#partitioned-topic) for the
public key of the node, encrypted with this key.

The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):

```protobuf

message PushNotificationOptions {
  boolean enabled = 1;
  repeated string allowed_user_list = 2;

>   repeated string block_user_list = 3;

Reviewer 2: Can we call this `blocked_user_list` to be consistent with `allowed_user_list`?

>   repeated string block_chat_list = 4;

Reviewer 2: Can we call this `blocked_chat_list` to be consistent with `allowed_user_list`?

}

message PushNotificationDeviceToken {
  enum TokenType {
    UNKNOWN_TOKEN_TYPE = 0;
    APN_TOKEN = 1;
    FIREBASE_TOKEN = 2;
  }
  TokenType token_type = 1;
  string token = 2;
  string installation_id = 3;
  PushNotificationPreferences preferences = 4;
}

message PushNotificationPreferences {
  repeated DeviceToken device_tokens = 1;
  uint version = 2;
  boolean unregister = 3;
}

message PushNotificationRegister {
  bytes payload = 1;
  bytes signature = 2;
}
```

A push notification server will handle the message according to the following rules:

- it MUST verify that the `signature` matches the public key of the sender.
- it MUST verify that `token_type` is supported
- it MUST verify that `token` is non empty
- it MUST verify that `installation_id` is non empty

> - it MUST verify that `version` is non-zero and greater than the currently stored, if any

Reviewer 2: "greater than the currently stored version for the public key of the sender" I suppose?

- it MUST verify that `device_tokens` is non empty

If `signature` does not match the public key of the sender, the message MUST be discarded.

If `token_type` is not supported, a response MUST be sent with `error` set to 
`UNSUPPORTED_TOKEN_TYPE`.

If `token`,`installation_id`,`device_tokens`,`version` are empty, a response MUST
be sent with `error` set to `MALFORMED_MESSAGE`.

If the `version` is equal or less than the currently stored version, a response MUST
be sent with `error` set to `VERSION_MISMATCH`.

If any other error occurs the `error` should be set to `INTERNAL_ERROR`.

If the request is successful, an access token MUST be generated and sent back to the client
with `success` set to `true`.

Otherwise a response MUST be sent with `success` set to `false`.

> It is RECOMMENDED to use a randomly generated string for `access_token` of at least 36 characters.

Reviewer 2: Where do we get this requirement of `at least 36 characters` from?

Author: It's a recommendation, not a requirement, but it's basically the length of an UUID, which looked safe enough as a secret

`request_id` should be set to the `SHA3-256` of the signature sent by the client.

> The request MUST be sent on the [partitioned topic][./10-waku-usage.md#partitioned-topic] of the sender.

Reviewer 2: ~request~ response

The payload of the response is:

```protobuf

> message PushNotificationPreferencesResponse {

Reviewer 2: Would this be better named as `PushNotificationRegistrationResponse` because the response is to the registration request?

  boolean success = 1;
  ErrorType error = 2;
  bytes request_id = 3;
  PushNotificationPreferences preferences = 4;
  string access_token = 5;

  enum ErrorType {
    UNKNOWN_ERROR_TYPE = 0;
    MALFORMED_MESSAGE = 1;
    VERSION_MISMATCH = 2;
    UNSUPPORTED_TOKEN_TYPE = 3;
    INTERNAL_ERROR = 4;
  }
}
```

[//]: (We can ratched and use a partitioned topic here)

A client SHOULD listen for a response sent on their [partitioned topic][./10-waku-usage.md#partitioned-topic].

If `success` is `true` the client has registered successfully.

If `success` is `false`:

- If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried without ensuring 

>   that is correctly formed.

Reviewer 2: retried ~without~ by ensuring that it is correctly formed.

Author: Not sure how it reads this one:
` If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried by ensuring 
  that is correctly formed.`

Did you mean also to change `SHOULD not` (maybe not should be capitalized), to `SHOULD`? In which case the meaning is slightly different, as I would like to make clear that the request SHOULD NOT be retried without modifying it, but the client might not retry it

Reviewer 2: Ah right. "If MALFORMED_MESSAGE is returned, the request SHOULD NOT be retried without ensuring that it is correctly formed" makes sense.

- If `INTERNAL_ERROR` is returned, the request MAY be retried, but the client MUST be 
  backoff exponentially

> - If `VERSION_MISMATCH` is returned, the client SHOULD ensure that the remote version returned in `preferences` is integrated, and the version should be incremented and the request MAY be sent again.

Reviewer 2: Change to "..the client SHOULD ensure that the version is incremented to beyond the remote version returned in `preferences` and then the request MAY be sent again."?

A client MAY register with multiple Push Notification Servers in order to increase availability.

A client SHOULD make sure that all the notification services they registered with have the same information about their devices and tokens.

If no response is returned the request SHOULD be considered failed and MAY be retried with the same server or a different one, but clients MUST exponentially backoff after each trial.

If the request is successful a token is returned. This SHOULD be 

> [advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.

Reviewer 2: This "public key filtering" refers to the allowed/blocked lists, correct?

Author: correct

## Re-registering with the push notification server

> A client SHOULD re-register with the node if the APN or FIREBASE token changes.

Reviewer 2: When does this happen? I suppose such a change is triggered by the client itself?

Author: Yes, it's up  to the client to refresh the token periodically with APN/Firebase

When re-registering a client SHOULD ensure that it has the most up-to-date
`PushNotificationPreferences`, update the part relative to their `installation_id`
if necessary, increment `version` and send a `PushNotificationRegister` as described above.

Once re-registered, a client SHOULD advertise the changes.

## Changing options

This is handled in exactly the same way as re-registering above.

## Unregistering from push notifications

To unregister a client MUST send a `PushNotificationRegister` request as described

> above with `unregister` in `PushNotificationPreferences` set to `true`, or removing

Reviewer 2: Why "or" removing their device information?

Author: In case you still want to keep some of your devices registered, basically you might want to unregister your android device, but not your IOS device (or you might want to unregister both, in which case the whole profile is removed).

their device information.

The server MUST remove all data about this user if `unregistering` is `true`,
apart from the `hash` of the public key and the `version` of the last options,
in order to make sure that old messages are not processed.

A client MAY unregister from a server on explicit logout if multiple accounts 

> are used on a single device.

Reviewer 2: Why is this `MAY` and not `SHOULD`? As a user who has logged out, I would not expect notifications meant for my chat key to be delivered to this device anymore, even if there are _no_ multiple accounts used on that device. That would be a privacy leak otherwise, wouldn't it?

Won't automatically registering (based on opt-in preferences) and unregistering on user login/logout be more in line with user's expectations?

Author: So I think this is a bit more of a UX concern,
but there are other things to consider that are implementation dependent, and I think `MAY` is best to give clients more freedom.

The best way to handle this is not to unregister only with the server on logout, because then you get into troubles if the user logs out when offline.

Say:
1) Go offline
2) Logout
3) Someone sends you a notification
4) You go online

The notification in 3 will be dispatched on IOS, there's no way around this as we don't have that kind of local control.

Ideally when you logout, you tell IOS that you don't want to receive notifications for that token, maybe by invalidating the previous one, but each client (android/IOS and maybe desktop) will kind of have to follow their own path. Some client might not even have a logout functionality (console client).

So I think MAY is more apt as it's very much related to the specific UX of a given client, but happy to change it to a SHOULD if you feel strongly about it.

Reviewer 2: Thanks for the clarification. Makes sense.

Every time a user register or re-register with a push notification service, their
contact-code SHOULD be re-advertised.

Multiple servers MAY be advertised for the same `installation_id` for redundancy reasons.

## Discovering a push notification server

[//]: We could query directly the nodes on a shared topic, but for simplicity and 
bandwidth usage this is the most convenient for now.

To discover a push notification service for a given user, their [contact code topic](./10-waku-usage.md#contact-code-topic)
SHOULD be listened to.
A mailserver can be queried for the specific topic to retrieve the most up-to-date
contact code.

## Querying the push notification server

If a token is not present in the latest advertisement for a user, the server 
SHOULD be queried directly.

To query a server a message:

```protobuf
message PushNotificationQuery {
  repeated bytes public_keys = 1;
}
```

MUST be sent to the server. 

The server MUST be checking the signature of the message and comparing
that with the latest `PushNotificationPreferences` for the given public key.

If the requester has access to any access token, a response MUST be sent:

```protobuf

> message PushNotificationQueryInfo {

Reviewer 2: This information should include the public key of user associated with the `access_token`, correct? Otherwise, how is a response bound to a certain query?

Author: yes, good point

  string access_token = 1;
  string installation_id = 2;
}

message PushNotificationQueryResponse {
  repeated PushNotificationQueryInfo info = 1;

}
```

Otherwise a response MUST NOT be sent.

Multiple queries MAY be sent for the same token from different public keys to
increase deniability.

A client MAY send request to the server using an ephemeral key at first, in order not to reveal 
their identity.

If the request is not successful, they should assume that some public key filtering is
in place at the server level and they MAY retry the request using their real identity.

## Sending a push notification

When sending a push notification, only the `installation_id` for the devices targeted 
by the message SHOULD be used.

If a message is for all the user devices, all the `installation_id` known to the client MAY be used. 

The number of devices MAY be capped in order to reduce resource consumption.

> At least 3 devices SHOULD be targeted, ordered by last activity.

Reviewer 2: Where does this requirement of `At least 3 devices` come from? Is it a heuristic?

Author: It's what we support by default in for messaging

- It should then [send](#sending-a-push-notification) a push notification according
  to the rules described
- The server should then send a request to the gorush server including all the required
  information

### Receiving a push notification

- On receiving the notification, a client can open the right account by checking the 
  `installation_id` included. The `chat_id` MAY be used to open the chat if present.
- `message` can be decrypted and presented to the user. Otherwise messages can be pulled from the mailserver if the `message_id` is no already present.

## Protobuf description

### PushNotificationRegister

A `PushNotificationRegister` is used to register with a Push Notification server.

`payload`: the protobuf encoded `PushNotificationPreferences`.

`signature`: the signature of the `payload` concatenated with the compressed `Secp256k` key of the Push Notification server. `Sig(PrivateKeyClient, payload + CompressedPublicKeyNode)`

#### Data disclosed

- Public key of the author

### PushNotificationPreferences

A push notification preferences message describes the push notification options and tokens for all the devices associated with `PublicKeyClient`.

> `device_tokens`: a list of `PushNotificationPreferences`, one for each device owned by the user.

Reviewer 2: Ah, so `device_tokens` are not really tokens but a list of `PushNotificationPreferences` which contain tokens among other things. This name was a bit confusing.

Author: Yes, should we rename it?

Reviewer 2: Better to IMO.

> `version`: an atomically increasing number identifying the current `PushNotificationPreferences`. Any time anything is changed in the record it MUST be increased by the client, otherwise the request will not be accepted.

Reviewer 2: ~atomically~ monotonically

`access_token`: the access token used to send a push notification 
`installation_id`: the `installation_id` of the device associated with the `access_token`

### PushNotificationQueryResponse

`info`:  a list of `PushNotificationQueryInfo`

### PushNotification

`access_token`: the access token used to send a push notification
`chat_id`: the `SHA3-256` of the `chat_id`

### Data disclosed

- The `SHA3-256` of the `chat_id` the notification is to be sent for

### PushNotificationRequest

`requests`: a list of `PushNotification`
`message`: the encrypted message that we want to notify on
`message_id`: the [status message id](./6-payloads.md)
`ack_required`: a 32 bytes long AES key, only set if an ack is required

### Data disclosed

- The status message id for which the notification is for
- the cypher text of the message

### PushNotificationAcknowledgement

`id`: the `ack_required` string passed by the client

## Partitioned topic

This is a modification of the [partitioned topic](./10-waku-usage.md#partitioned-topic) used by clients

```golang
var partitionsNum *big.Int = big.NewInt(5000)
var partition *big.Int = big.NewInt(0).Mod(publicKey.X, partitionsNum)

> partitionTopic := "push-notifications-" + strconv.FormatInt(partition.Int64(), 10)

Reviewer 2: So this prefix of `push-notifications-` is the only modification?

Author: correct

var hash []byte = keccak256(partitionTopic)
var topicLen int = 4

if len(hash) < topicLen {
    topicLen = len(hash)
}

var topic [4]byte
for i = 0; i < topicLen; i++ {
    topic[i] = hash[i]
}
```

## Anonymous mode of operations

An anonymous mode of operations MAY be provided by the client, where the 
responsibility of propagating information about the user is left to the client,
in order to preserve privacy.

A client in anonymous mode can register with the server using a key different 

> from their identity key.

Reviewer 2: What is this `identity key`? Is this the chat key?

Author: yes

Reviewer 2: Then should we call it `chat key` here unless `identity key` is the standard terminology used elsewhere?

This will hide their real identity key.

This identity key is effectively a secret and SHOULD only be disclosed to clients that you the user wants to be notified by.

A client MAY advertise the access token on the contact-code topic of the key generated.
A client MAY share their identity key through [contact updates](./6-payloads.md#contact-update)

A client receiving a push notification identity key SHOULD listen to the contact code 
topic of the push notification identity key for updates.

If a client has received a push notification identity key they MAY request 
the access token using an ephemeral key when querying the server.

The method described above effectively does not share the identity of the sender
nor the receiver to the server, but MAY result in missing push notifications as
the propagation of the secret is left to the client.

This can be mitigated by [device syncing](./6-payloads.md), but not completely
addressed.

> [//]: This section can be removed, for now leaving it here in order to help with the 

Reviewer 2: Sorry, I don't fully understand how this works yet.

Author: The  whole thing or only the device syncing part? I can further elaborate if necessary

Reviewer 2: Ok, I think I understand this now (except device syncing). The terminology (specifically, use of the words identity key, push notification identity key and secret :) threw me off earlier, sorry. 

So the difference is in the generation of a push notification key (PN key) which is different from the user's chat key (i.e. identity key). The msg sender has to be aware of this PN key to send push notifications to the recipient and the probabilistic (?) propagation of this PN key makes the receipt of PNs indeterministic. 

The PN sender may get the access token either by querying the PN server like earlier (using an ephemeral key if it wants) or directly receive it from the client by listening to contact-code topic for the PN key. The PN sender then sends a notification just like earlier by talking to the PN server which in turn talks to the gorush server.

The use of ephemeral PN-specific keys makes this mode anonymous.

Author: Correct,
the device syncing part is that if you have multiple devices, say A1 & A2, and you chose to propagate the token yourself (so not through the server), you need to make sure that A1 & A2 are paired, otherwise they might not be aware of each other, and therefore you might be propagating only the token for A1 to another client for example, which will only send a PN to A1.

  - [Protobuf description](#protobuf-description)
    - [PushNotificationRegister](#pushnotificationregister)
    - [PushNotificationPreferences](#pushnotificationpreferences)
    - [PushNotificationDeviceToken](#pushnotificationdevicetoken)
    - [PushNotificationOptions](#pushnotificationoptions)
    - [PushNotificationPreferencesResponse](#pushnotificationpreferencesresponse)
    - [ContactCodeAdvertisement](#contactcodeadvertisement)
    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
    - [PushNotificationQuery](#pushnotificationquery)
    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
    - [PushNotification](#pushnotification)
    - [PushNotificationRequest](#pushnotificationrequest)
    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
  - [Partitioned topic](#partitioned-topic)
  - [Anonymous mode of operations](#anonymous-mode-of-operations)
  - [Security considerations](#security-considerations)
  - [FAQ](#faq)
  - [Changelog](#changelog)
    - [Version 0.1](#version-01)
  - [Copyright](#copyright)

## Reason

> Push notifications for IOS devices and for some android device can only be

Reviewer 3: ```suggestion
Push notifications for iOS devices and for some Android devices can only be
```

> implemented relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for IOS or [Firebase](https://firebase.google.com/).

Reviewer 3: ```suggestion
implemented by relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for iOS or [Firebase](https://firebase.google.com/).
```

This is useful for Android devices that do not support background services or that often kill the 
background service.

iOS only allows certain kind of applications to keep a connection open when in the
background, VoIP for example, which current status client does not qualify for.

Applications on iOS can also request execution time when they are in the [background]

> (https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)

Reviewer 4: Link doesn't show correctly

```suggestion
Applications on iOS can also request execution time when they are in the [background](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)
```

> Applications on iOS can also request execution time when they are in the [background]
> (https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)

Reviewer 3: ```suggestion
Applications on iOS can also request execution time when they are in the [background](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)
```

MUST be given an option not to enable push notifications and not to send push notifications.

## Requirements

The party releasing the app needs to run their own [gorush](https://github.com/appleboy/gorush) 
publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.

## Components

### Gorush instance

A [gorush](https://github.com/appleboy/gorush) instance MUST be publicly 
available, this will be used only by push notification servers.

### Push notification server

A push notification server used by clients to register for push notification and
sending push notifications.

### Registering client

A status client that wants to receive push notifications

### Sending client

A status client that wants to send push notifications

## Registering with the push notification service

A client MAY register with a Push Notification service of their choice.

[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)

A message MUST be sent to the [partitioned topic](#partitioned-topic) for the

> public key of the node, encrypted with this key.

Reviewer 4: This is a special notification registration message, perhaps we should give this message a special dedicated name. Something like:

```suggestion
A `PNR message` (Push Notification Registration) MUST be sent to the [partitioned topic](#partitioned-topic) for the
public key of the node, encrypted with this key.
```

The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):

```protobuf

message PushNotificationOptions {
  boolean enabled = 1;
  repeated string allowed_user_list = 2;
  repeated string block_user_list = 3;
  repeated string block_chat_list = 4;
}

message PushNotificationDeviceToken {
  enum TokenType {
    UNKNOWN_TOKEN_TYPE = 0;
    APN_TOKEN = 1;
    FIREBASE_TOKEN = 2;
  }
  TokenType token_type = 1;
  string token = 2;
  string installation_id = 3;
  PushNotificationPreferences preferences = 4;
}

message PushNotificationPreferences {
  repeated DeviceToken device_tokens = 1;
  uint version = 2;
  boolean unregister = 3;
}

> message PushNotificationRegister {

Reviewer 5: How do we prevent this message to be intercepted and replayed to sign up for other servers? An attacker could do so to compromise privacy, spam notifications and if a cost is attached drain user funds

Author: Good question, so:

1) the message is sent to the server encrypted with their public key, so only the server is able to  access it. 

2) If replied to the same server the actions taken are idempotent.
3)  If you look at `PushNotificationRegister` definition, we `Sign(payload+publickeyoftheserver)` so that any server receiving this message will discard it if not for them.

Reviewer 5: Okay so with these assumptions replaying is only possible if an eavesdropper waits for a user to register and register.

To me the question remains if a server would still discard a notifications sign up if a payment is involved and potentially claimable with this message.

Author: > Okay so with these assumptions replaying is only possible if an eavesdropper waits for a user to register and register.

Not sure I follow,
you can only reply this to the same server, as the signature would fail to verify and it would be discarded. 
Replying to the same server will not yield to any result, as the message would be discarded if already processed (it's basically idempotent), due to a `version` inside the message that is part of the hash to be signed, which effectively acts as a `nonce`.

Given the above, is this still replayable in any meaningful sense?

> To me the question remains if a server would still discard a notifications sign up if a payment is involved and potentially claimable with this message.

What do you mean exactly? You mean that someone could offer money to a malicious server to drop the message/reveal the message?

Reviewer 5: Could you elaborate on how the version could replace a nonce? I was thinking about case: user registers, user unregisters, attacker replays to register again.

For the case of malicious servers I was referring to the notifications as a service case (I didn't find specs for that but iirc it was somehow a longterm plan) , depending on the implementation, if the payment is part of the message payload a replay to a malicious server could be used to steal user funds

Author: >  I was thinking about case: user registers, user unregisters, attacker replays to register again.

In this case it won't work, `version` is basically the same as an ethereum nonce (minus the fact that they need to be strictly incrementing by one), basically the server will only process requests that have a version strictly greater than the previous one.
So in your case, user will register with a version of `1`. They will unregister, as per spec this will need to have a version of `2`. The server will keep the last version (`2`) stored, as per spec:
```
The server MUST remove all data about this user if `unregistering` is `true`,
apart from the `hash` of the public key and the `version` of the last options,
in order to make sure that old messages are not processed.
```
If `1` (or `2`) is replayed, the server will ignore the message, as the `version` is not strictly greater than `2`

Reviewer 2: Why use `version` and not `nonce`?

Author: I think this is more of a `version` than a `nonce`, a `nonce` is literally a `number used once`.
In our case 1) we always refer to the same data-structure, which clients are supposed to pull from the server and update 2) The nonce needs to be monotonically increasing 

I guess it depends on what people are most familiar with, but this is an established pattern by distributed databases using optimistic locking and generally version is used ( elasticsearch/dynamodb)

- it MUST verify that `token_type` is supported
- it MUST verify that `token` is non empty
- it MUST verify that `installation_id` is non empty
- it MUST verify that `version` is non-zero and greater than the currently stored, if any
- it MUST verify that `device_tokens` is non empty

If `signature` does not match the public key of the sender, the message MUST be discarded.

If `token_type` is not supported, a response MUST be sent with `error` set to 
`UNSUPPORTED_TOKEN_TYPE`.

If `token`,`installation_id`,`device_tokens`,`version` are empty, a response MUST
be sent with `error` set to `MALFORMED_MESSAGE`.

If the `version` is equal or less than the currently stored version, a response MUST
be sent with `error` set to `VERSION_MISMATCH`.

If any other error occurs the `error` should be set to `INTERNAL_ERROR`.

If the request is successful, an access token MUST be generated and sent back to the client
with `success` set to `true`.

Otherwise a response MUST be sent with `success` set to `false`.

It is RECOMMENDED to use a randomly generated string for `access_token` of at least 36 characters.

`request_id` should be set to the `SHA3-256` of the signature sent by the client.

> The request MUST be sent on the [partitioned topic][./2-account.md] of the sender.

Reviewer 4: Should this be `[partitioned topic](#partitioned-topic)`? I can't see any reference to partitioned topic in the account specs.

Author: It should be `10-waku-usage.md#partitioned-topic` good catch

The payload of the response is:

```protobuf
message PushNotificationPreferencesResponse {
  boolean success = 1;
  ErrorType error = 2;
  bytes request_id = 3;
  PushNotificationPreferences preferences = 4;
  string access_token = 5;

  enum ErrorType {
    UNKNOWN_ERROR_TYPE = 0;
    MALFORMED_MESSAGE = 1;
    VERSION_MISMATCH = 2;
    UNSUPPORTED_TOKEN_TYPE = 3;
    INTERNAL_ERROR = 4;
  }
}
```

[//]: (We can ratched and use a partitioned topic here)

> A client SHOULD listen for a response sent on their [partitioned topic][./2-account.md].

Reviewer 4: Same as above

If `success` is `true` the client has registered successfully.

If `success` is `false`:

- If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried without ensuring 
  that is correctly formed.
- If `INTERNAL_ERROR` is returned, the request MAY be retried, but the client MUST be 
  backoff exponentially
- If `VERSION_MISMATCH` is returned, the client SHOULD ensure that the remote version returned in `preferences` is integrated, and the version should be incremented and the request MAY be sent again.

A client MAY register with multiple Push Notification Servers in order to increase availability.

A client SHOULD make sure that all the notification services they registered with have the same information about their devices and tokens.

> If no response is returned the request SHOULD be considered failed and MAY be retried with the same server or a different one, but it MUST exponentially backoff after each trial.

Reviewer 4: >but it MUST exponentially backoff after each trial

Does "it" refer to a client?

Author: Yes, I'll use `clients` instead

If the request is successful a token is returned. This SHOULD be 
[advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.

## Re-registering with the push notification server

A client SHOULD re-register with the node if the APN or FIREBASE token changes.

When re-registering a client SHOULD ensure that it has the most up-to-date
`PushNotificationPreferences`, update the part relative to their `installation_id`
if necessary, increment `version` and send a `PushNotificationRegister` as described above.

> Once re-registered, a client SHOULD advertise the changes.

Reviewer 4: Does the node need special behaviour for re-registering, or will the previously given registration details be overwritten and behave exactly the same as a standard registration?

Author: No special behavior, the previously given registration details will be overwritten.

Reviewer 2: So for the PN server, this will look like a new registration request except for version check?

Author: Correct

}
```

If no filtering is done based on public keys,
the access token SHOULD be included in the advertisement.
Otherwise it SHOULD be left empty.

This SHOULD be advertised on the [contact code topic](./10-waku-usage.md#contact-code-topic)
and SHOULD be coupled with normal contact-code advertisement.

Every time a user register or re-register with a push notification service, their
contact-code SHOULD be re-advertised.

Multiple servers MAY be advertised for the same `installation_id` for redundancy reasons.

## Discovering a push notification server

[//]: We could query directly the nodes on a shared topic, but for simplicity and 
bandwidth usage this is the most convenient for now.

To discover a push notification service for a given user, their [contact code topic](./10-waku-usage.md#contact-code-topic)
SHOULD be listened to.
A mailserver can be queried for the specific topic to retrieve the most up-to-date
contact code.

## Querying the push notification server

If a token is not present in the latest advertisement for a user, the server 
SHOULD be queried directly.

To query a server a message:

```protobuf
message PushNotificationQuery {
  repeated bytes public_keys = 1;

> }

Reviewer 5: Same as above, we probably need a nonce or this message might be replayed to get access to the tokens

Author: mmh, not sure, first the message is encrypted with the server pk, so to replay this you would need to have access to the private key of the server.

Say that this happen and the key is compromised,an attacker would yes be able to replay the same message, but it would not be able to listen to the reply, as it's encrypted with the pk of the user querying.

If the user has access to both, then it would not make sense to replay a message, as it can just craft a new one.

I mean, we can add a nonce, but in order to replay this effectively you would have to compromise two private keys (or one, the user querying, but at that point you can craft a new message, listen to replies etc), and this message does not change any state, so you would only replay it to get access to information.

Or you think there is still some reason to add a nonce?

A client MAY send request to the server using an ephemeral key at first, in order not to reveal 
their identity.

If the request is not successful, they should assume that some public key filtering is
in place at the server level and they MAY retry the request using their real identity.

## Sending a push notification

When sending a push notification, only the `installation_id` for the devices targeted 
by the message SHOULD be used.

If a message is for all the user devices, all the `installation_id` known to the client MAY be used. 

The number of devices MAY be capped in order to reduce resource consumption.

At least 3 devices SHOULD be targeted, ordered by last activity.

For any device that a token is available, or that a token is successfully queried,
a push notification message SHOULD be sent to the corresponding push notification server.

```protobuf
message PushNotification {
  string access_token = 1;
  string chat_id = 2;
}

message PushNotificationRequest {
  repeated PushNotification requests = 1;
  bytes message = 2;
  string message_id = 3;
  string ack_required = 4;
}
```

Where `message` is the encrypted payload of the message and `chat_id` is the

> `SHA3-256` of the `chat_id`.

Reviewer 5: Why do we hash here? Is chatid random and the preimage space large enough?

Author: `ChatID` is random in some cases (group chats), and not random in some other (one to one) where is the public key of the user.
Preimage is large enough in both cases.

A client MAY advertise the access token on the contact-code topic of the key generated.
A client MAY share their identity key through [contact updates](./6-payloads.md#contact-update)

A client receiving a push notification identity key SHOULD listen to the contact code 
topic of the push notification identity key for updates.

If a client has received a push notification identity key they MAY request 
the access token using an ephemeral key when querying the server.

The method described above effectively does not share the identity of the sender
nor the receiver to the server, but MAY result in missing push notifications as
the propagation of the secret is left to the client.

This can be mitigated by [device syncing](./6-payloads.md), but not completely
addressed.

[//]: This section can be removed, for now leaving it here in order to help with the 
review process. Point can be integrated, suggestion welcome.

## Security considerations

If no anonymous mode is used, when registering with a push notification service a client discloses:

- Their public key
- Which devices they want to receive notifications on

> They MAY disclose:

Reviewer 4: Does "They" refer to a client?

Author: yes, changed to avoid possessive adjectives

multiple users are using the same push notification server. Queries have also 
a bandwidth impact but they are made only when actually needed

### What's the information disclosed?

The data disclosed with each message sent by the client is above, but for a summary:

When you register with a push notification service you may disclose:

1) Your public key
2) Which devices you have
3) The hash of the chat_ids you want to filter out
4) The has of the public keys you are interested/not interested in

When you query a notification service you may disclose:

1) Your public key
2) The fact that you are interested in sending push notification to a given user

Effectively this is fairly revealing if the user has a whitelist implemented.
Therefore sending notification should be optional.

### What prevents a user from generating a random key and getting an access token and spamming?

Nothing really, that's the same as the status app as a whole. the only mechanism that prevents
this is using a white-list as described above, but that implies disclosing your true identity to the push notification server.

### Why not 0-knowledge proofs/quantum computing

> We start simple, we can iterate

Reviewer 4: Beautifully succinct

the advertisement as that's effectively a broadcast, but those info should not change and we can
always accrete the message.

### Why ack_required?

That's necessary to avoid duplicated push notifications.

Deduplication of the push notification is done on the client side, to relive a bit 
of centralization and also in order not to have to modify gorush. 

### Can I run my own node?

Sure, the methods allow that

### Can I register with multiple nodes for redundancy

Yep

### What does my node disclose?

Your node will disclose the IP address is running from, as it makes an HTTP post to 
gorush. A waku adapter could be used, but please not now.

### Does this have high-reliability requirements?

The gorush server yes, no way around it.

The rest, kind of, at least one node having your token needs to be up for you to receive notifications.
But you can register with multiple servers (desktop, status, etc) if that's a concern.

### Can someone else (i.e not status) run this?

Push notification servers can be run by anyone. Gorush can be run by anyone I take,
but we are in charge of the certificate, so they would not be able to notify status-clients.

## Changelog

### Version 0.1

> Released [May 22, 2020](https://github.com/status-im/specs/commit/664dd1c9df6ad409e4c007fefc8c8945b8d324e8)

Reviewer 4: Is this true? I thought that this spec was only drafted this month. Maybe just a copypasta from another file?

Author: No, that is indeed a copypasta, I will fill it in once ready to merge :)
